File: app/src/main/java/ga/forntoh/bableschool/fragments/TermScoreFragment.java
package ga.forntoh.bableschool.fragments;


import android.annotation.SuppressLint;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.forntoh.EasyRecyclerView.EasyRecyclerView;
import com.google.gson.reflect.TypeToken;
import com.nytimes.android.external.store3.base.impl.Store;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Objects;

import ga.forntoh.bableschool.R;
import ga.forntoh.bableschool.adapters.TermScoresAdapter;
import ga.forntoh.bableschool.model.Score;
import ga.forntoh.bableschool.network.ApiService;
import ga.forntoh.bableschool.network.RetrofitBuilder;
import ga.forntoh.bableschool.store.MyStores;
import ga.forntoh.bableschool.store.StorageUtil;
import ga.forntoh.bableschool.store.StoreRepository;
import ga.forntoh.bableschool.utils.Utils;
import io.reactivex.schedulers.Schedulers;

/**
 * A simple {@link Fragment} subclass.
 */
@SuppressWarnings("unchecked")
public class TermScoreFragment extends Fragment {


    private boolean fetched = false;
    private ArrayList<Score> scores = new ArrayList<>();
    private TermScoresAdapter termScoresAdapter;

    public TermScoreFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_term_score, container, false);

        if (getArguments() != null) {

            termScoresAdapter = new TermScoresAdapter(scores);
            new EasyRecyclerView()
                    .setType(EasyRecyclerView.Type.VERTICAL)
                    .setAdapter(termScoresAdapter)
                    .setRecyclerView(v.findViewById(R.id.rv_term_scores))
                    .setItemSpacing(16, null)
                    .initialize();

            fetchItems(getArguments().getInt("term"));
        }

        return v;
    }

    @SuppressLint("CheckResult")
    private void fetchItems(int term) {
        ApiService service = RetrofitBuilder.createService(ApiService.class);
        StoreRepository repository = new StoreRepository(Objects.requireNonNull(getActivity()).getApplication());

        Store store = repository
                .create(service.getTermScores(StorageUtil.with(getActivity()).loadMatriculation(), term, Utils.getTermYear()), new TypeToken<List<Score>>() {
                }.getType());

        if (fetched)
            store
                    .get(MyStores.TERM_SCORE(getContext(), term))
                    .subscribeOn(Schedulers.io())
                    .subscribe(this::dealWithScores, t -> ((Throwable) t).printStackTrace());
        else {
            store
                    .fetch(MyStores.TERM_SCORE(getContext(), term))
                    .subscribeOn(Schedulers.io())
                    .subscribe(scores -> {
                        this.dealWithScores(scores);
                        fetched = true;
                    }, t -> ((Throwable) t).printStackTrace());
        }
    }

    @SuppressWarnings("ConstantConditions")
    private void dealWithScores(Object o) {
        this.scores.clear();
        this.scores.addAll((Collection<? extends Score>) o);
        this.scores.add(0, null);
        getActivity().runOnUiThread(() -> termScoresAdapter.notifyDataSetChanged());
    }

}

File: app/src/main/java/ga/forntoh/bableschool/fragments/TermScoreFragment.kt
package ga.forntoh.bableschool.fragments

import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.forntoh.EasyRecyclerView.EasyRecyclerView
import com.raizlabs.android.dbflow.kotlinextensions.*
import ga.forntoh.bableschool.ApiService
import ga.forntoh.bableschool.R
import ga.forntoh.bableschool.RetrofitBuilder
import ga.forntoh.bableschool.StorageUtil
import ga.forntoh.bableschool.adapters.TermScoresAdapter
import ga.forntoh.bableschool.model.Score
import ga.forntoh.bableschool.model.Score_Table
import ga.forntoh.bableschool.utils.Utils
import ga.forntoh.bableschool.utils.Utils.dealWithData
import ga.forntoh.bableschool.utils.Utils.isConnected
import io.reactivex.schedulers.Schedulers

class TermScoreFragment : Fragment() {

    private val scores = ArrayList<Score>()
    private val termScoresAdapter by lazy { TermScoresAdapter(scores) }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val v = inflater.inflate(R.layout.fragment_term_score, container, false)
        if (arguments != null) {
            EasyRecyclerView()
                    .setType(EasyRecyclerView.Type.VERTICAL)
                    .setAdapter(termScoresAdapter)
                    .setRecyclerView(v.findViewById(R.id.rv_term_scores))
                    .setItemSpacing(16, null)
                    .initialize()
            fetchItems(arguments!!.getInt("term"))
        }
        return v
    }

    private fun fetchItems(term: Int) {
        val service = RetrofitBuilder.createService(ApiService::class.java)
        if (isConnected(activity!!))
            service.getTermScores(StorageUtil.getInstance(activity!!.baseContext).loadMatriculation(), term, Utils.termYear)
                    .subscribeOn(Schedulers.io())
                    .subscribe({ scores ->
                        (delete(Score::class) where Score_Table.term.eq(term)).execute()
                        if (!scores.isNullOrEmpty()) scores.forEach { it.term = term; it.save() }
                        dealWithData(activity!!, (select from Score::class where Score_Table.term.eq(term)).list.apply { add(0, Score()) }, this.scores, termScoresAdapter)
                    }) { dealWithData(activity!!, (select from Score::class where Score_Table.term.eq(term)).list.apply { add(0, Score()) }, this.scores, termScoresAdapter) }
        else {
            dealWithData(activity!!, (select from Score::class where Score_Table.term.eq(term)).list.apply { add(0, Score()) }, this.scores, termScoresAdapter)
        }
    }
}
